File: static-src/textsearch/fold.ts
/**
 * Lowercases for a case-insensitive scan with Unicode's simple, per-code-point
 * mapping (no Final_Sigma, no locale): U+0130 becomes `i`, then every code point
 * is lowercased on its own. Length-preserving in UTF-16 code units by
 * construction, `fold(s).length === s.length`, always, so an index into fold(s)
 * is an index into s. A lone surrogate is kept as it is.
 */
export function fold(s: string): string {
  let out = "";
  for (const cp of s) {
    out += cp === "\u0130" ? "i" : cp.toLowerCase();
  }
  return out;
}

File: internal/textsearch/fold.go
package textsearch

import (
	"strings"
	"unicode"
)

// Fold lowercases for a case-insensitive scan with Unicode's simple, per-rune
// mapping (no Final_Sigma, no locale). Rune-count-preserving by construction:
// strings.Map with a rune-to-rune function, an invalid byte becoming one
// U+FFFD. So a rune index into Fold(s) is a rune index into s. Byte offsets
// are NOT preserved; Occurrences reports the original's.
func Fold(s string) string {
	return strings.Map(unicode.ToLower, s)
}
